Events Dashboard

michael07w opened the pull request:
Method returns events that a particular user is a moderator of, or has been invited to.

michael07w committed:
feat(server): added method to return events that a user is a moderator of or has been invited t

michael07w committed:
feat(server): added method to return events that a user is a moderator of or has been invited t

michael07w committed:
feat(server): added method to return events that a user is a moderator of or has been invited t

michael07w committed:
feat(server): added resolver to return a particular user's events

Updated graphql schema to support query for a User's events. Also, added resolver to return proper events. Tested with Postman to confirm query returns expected results.

michael07w committed:
feat(server): added resolver to return a particular user's events

Updated graphql schema to support query for a User's events. Also, added resolver to return proper events. Tested with Postman to confirm query returns expected results.

michael07w committed:
feat(server): added resolver to return a particular user's events

Updated graphql schema to support query for a User's events. Also, added resolver to return proper events. Tested with Postman to confirm query returns expected results.

michael07w committed:
feat(server): added resolver to return a particular user's events

Updated graphql schema to support query for a User's events. Also, added resolver to return proper events. Tested with Postman to confirm query returns expected results.

michael07w committed:
feat(frontend): built queries/fragments to display necessary events in dashboard

michael07w committed:
feat(frontend): built queries/fragments to display necessary events in dashboard

michael07w committed:
feat(frontend): built queries/fragments to display necessary events in dashboard

michael07w committed:
feat(frontend): built queries/fragments to display necessary events in dashboard

michael07w committed:
feat(frontend): Implemented sorting of events into current and upcoming

michael07w committed:
feat(frontend): Implemented sorting of events into current and upcoming

michael07w committed:
feat(frontend): Implemented sorting of events into current and upcoming

michael07w committed:
feat(frontend): Implemented sorting of events into current and upcoming

michael07w committed:
feat(frontend): Updated event layout to match the format seen in 'My Organizations' page
:skip
:skip

michael07w committed:
feat(frontend): Updated event layout to match the format seen in 'My Organizations' page
:skip
:skip

michael07w committed:
feat(frontend): Updated event layout to match the format seen in 'My Organizations' page
:skip
:skip

michael07w committed:
feat(frontend): Added organization name; Updated event sorting logic
Added organization name to each Dashboard event item; Updated event sorting logic to separate current and upcoming events
:skip

michael07w committed:
feat(frontend): Added organization name; Updated event sorting logic
Added organization name to each Dashboard event item; Updated event sorting logic to separate current and upcoming events
:skip

michael07w committed:
feat(frontend): Added organization name; Updated event sorting logic
Added organization name to each Dashboard event item; Updated event sorting logic to separate current and upcoming events
:skip

kadey001 commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,232 @@
+            const eventEndDate = new Date(event.endDateTime.toString());
+            const eventStartTime = eventStartDate.getTime();
+            const eventEndTime = eventEndDate.getTime();
+            if (currentDate.year === eventStartDate.getFullYear() && 
```
I don't think the year/month/day checks are necessary if you just compare date.getTime() you can see how it works in the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)

kadey001 commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,232 @@
+    if(!data) return <Loader />;
+
+    // Get current date to categorize events as current or upcoming
+    const d = new Date();
```
you should use more descriptive variable names for this like currentDate for example.

kadey001 commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,232 @@
+
+    // Store current events
+    // Would prefer to use `isActive` member of Event node, but this doesn't seem to have been implemented yet
+    const current_events = listOfEvents.filter(({ node: event }) => {
```
We want to keep variable [naming conventions](https://curc.readthedocs.io/en/latest/programming/coding-best-practices.html) consistent so you should use Camelcase here rather than snake case.

kadey001 commented on app/client/src/features/dashboard/DashboardEvent.tsx:
```diff
@@ -0,0 +1,56 @@
+import { ListItemText } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+
+// import TitleCard from '@local/components/TitleCard';
+// import FadeThrough from '@local/animations/FadeThrough';
+// import RequireRoles from '@local/domains/Logical/RequireRoles';
+// import RoleInvite from '@local//Admin/RoleInvite';
```
Remove unused imports unless you plan to uncomment these later.

michael07w committed:
feat(frontend): Updated Date comparisons, renamed variables, and added live video link
:skip
:skip

michael07w committed:
feat(frontend): Updated Date comparisons, renamed variables, and added live video link
:skip
:skip

michael07w committed:
feat(frontend): Updated Date comparisons, renamed variables, and added live video link
:skip
:skip

michael07w committed:
feat(frontend): Add missing semicolons
:skip
:skip

michael07w committed:
feat(frontend): Add missing semicolons
:skip
:skip

michael07w committed:
feat(frontend): Add missing semicolons
:skip
:skip

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,218 @@
+
+const useStyles = makeStyles((theme) => ({
+    item: {
+        margin: theme.spacing(0, 0, 4, 0),
```
This could be `marginBottom: theme.spacing(4)`

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,218 @@
+        margin: theme.spacing(0, 0, 4, 0),
+    },
+    card: {
+        padding: theme.spacing(1, 1, 1, 1),
```
This could be `padding: theme.spacing(1)`

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,218 @@
+    });
+
+    return (
+        // <FadeThrough animKey='dashboard-page'>
```
Remove if unused.

michael07w committed:
fix(frontend): Updated material-ui imports to match upgrade to v5
:skip
:skip

michael07w committed:
fix(frontend): Updated material-ui imports to match upgrade to v5
:skip
:skip

michael07w committed:
fix(frontend): Updated material-ui imports to match upgrade to v5
:skip
:skip

michael07w committed:
fix(frontend): more updates to material-ui imports for transition to v5

michael07w committed:
fix(frontend): more updates to material-ui imports for transition to v5

michael07w committed:
fix(frontend): more updates to material-ui imports for transition to v5

michael07w committed:
test(e2e): Implemented auto-authentication for testing. Also implemented dashboard playwright tests.

michael07w committed:
test(e2e): Implemented auto-authentication for testing. Also implemented dashboard playwright tests.

michael07w committed:
test(e2e): Implemented auto-authentication for testing. Also implemented dashboard playwright tests.

michael07w committed:
test(e2e): Implemented auto-authentication for testing. Also implemented dashboard playwright tests.

michael07w committed:
feat(frontend): Implemented redirect to landing page, if user is not logged in.

michael07w committed:
feat(frontend): Implemented redirect to landing page, if user is not logged in.

michael07w committed:
feat(frontend): Implemented redirect to landing page, if user is not logged in.

michael07w committed:
test(dashboard): Implemented dashboard test specific to mobile.

michael07w committed:
test(dashboard): Implemented dashboard test specific to mobile.

michael07w committed:
test(dashboard): Implemented dashboard test specific to mobile.

michael07w committed:
feat(frontend): Wrapped Live Feed button in Link component to avoid redirect from event settings page to live page. The ListItem was directing user to event settings, when clicking the Live Feed button, before redirecting to the live page.

michael07w committed:
feat(frontend): Wrapped Live Feed button in Link component to avoid redirect from event settings page to live page. The ListItem was directing user to event settings, when clicking the Live Feed button, before redirecting to the live page.

michael07w committed:
feat(frontend): Wrapped Live Feed button in Link component to avoid redirect from event settings page to live page. The ListItem was directing user to event settings, when clicking the Live Feed button, before redirecting to the live page.

michael07w committed:
test(dashboard): Implemented all e2e testing for Dashboard.

michael07w committed:
test(dashboard): Implemented all e2e testing for Dashboard.

michael07w committed:
test(dashboard): Implemented all e2e testing for Dashboard.

michael07w committed:
test(dashboard): Implemented support for certain tests to run w/o authentication

michael07w committed:
test(dashboard): Implemented support for certain tests to run w/o authentication

michael07w committed:
test(dashboard): Implemented support for certain tests to run w/o authentication

michael07w commented:
All tests have passed. Ready for Review. #315 
![image](https://user-images.githubusercontent.com/29931619/174917679-8da49c1f-220b-4fc0-9558-722ae8d8a642.png)

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,224 @@
+    // Get current date to categorize events as current or upcoming
+    const currentDateTime = new Date();
+
+    // Store current events
+    const currentEvents = listOfEvents.filter(({ node: event }) => {
+        if (event.startDateTime && event.endDateTime) {
+            const eventStartDateTime = new Date(event.startDateTime);
+            const eventEndDateTime = new Date(event.endDateTime);
+            if (eventStartDateTime.getTime() < currentDateTime.getTime() && 
+                eventEndDateTime.getTime() > currentDateTime.getTime()) {
+                return true;
+            }
+        }
+    });
```
This could be a little more succinct / clear.  I'd also rename to ongoing. "current" is a little more ambiguous than ongoing in this context imo.


Also, we could wrap this in a React.useMemo.
```tsx
import { isBefore, isAfter } from 'date-fns';
// ...
const ongoingEvents = React.useMemo(() => listOfEvents.filter(({ node: event }) => {
  if (!event.startDateTime || !event.endDateTime) return false; // don't include events without both a start and end date
  const now = new Date();
  return isBefore(new Date(event.endDateTime), now) && isAfter(new Date(event.startDateTime), now);
}), [listOfEvents]);
```

https://date-fns.org/v2.28.0/docs/isBefore
https://date-fns.org/v2.28.0/docs/isAfter
https://reactjs.org/docs/hooks-reference.html#usememo

DevelopmentByDavid replied:
Ah, one more thing, you'll probably get a lint error since this is technically after a return statement. Just move the return statement lower down. It's not that expensive to calculate these things on the initial render especially if there's no data and it's loading.

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,224 @@
+    if(!data) return <Loader />;
+
+    // Get current date to categorize events as current or upcoming
+    const currentDateTime = new Date();
```
I don't really think this is expensive enough to justify trying to cache this. Also, this will technically be re-instantiated each re-render.

I'd just declare inline as needed.

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,224 @@
+        if (!isLoading && !user) router.push('/');
+    }, [user, router, isLoading]);
+    
+    if(!data) return <Loader />;
```
Space is missing. Make sure to run prettier to format.

`if(...` -> `if (...`

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,224 @@
+
+    // Store upcoming events
+    const upcomingEvents = listOfEvents.filter(({ node: event }) => {
+        if (event.startDateTime) {
+            const eventStartDateTime = new Date(event.startDateTime);
+            if (currentDateTime < eventStartDateTime) {
+                return true;
+            }
+        }
```
In general, prefer doing whatever if statement/logic will reduce the amount of nesting. Makes the code look a bit cleaner IMO. 

Also, rather than doing `if (boolean condition) return true/false` it's a bit cleaner to do `return boolean condition` (if it's possible / there's no other branching logic.)
```ts
import { isBefore } from 'date-fns';
// ...
if (!event.startDateTime) return false;
return isBefore(new Date(event.startDateTime), new Date());
```

DevelopmentByDavid commented on app/client/src/features/dashboard/DashboardEvent.tsx:
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable react/jsx-props-no-spreading */
```
Is this necessary here? I don't see any props spreading.

DevelopmentByDavid commented on app/client/src/features/dashboard/DashboardEvent.tsx:
```diff
@@ -0,0 +1,29 @@
+    organization: string,
+}
+
+export function DashboardEvent(props: DashboardEventProps) {
```
I think this should be giving you a lint warning? I'm not actually 100% sure and it's not a big deal if you're not getting one. In most components, we destructure the props;

`export function DashboardEvent({ id, title, description, startDateTime, organization }: DashboardEventProps) {`

DevelopmentByDavid commented on app/client/storageState.json:
```diff
@@ -0,0 +1,15 @@
+    {
+      "sameSite": "Lax",
+      "name": "jwt",
+      "value": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IlZYTmxjanBpWlRVd056TTFaUzAxWldGakxUUTNPRGN0T0RjMVl5MHhOelU1TlRRNE9HVTJaV0k9IiwiaWF0IjoxNjU1MTY3NDkxfQ.dywfF4Q--mbFf875TRbijy3PB0wRX6kdOpfYS736dqA",
```
Could you add a page to the docs on what this should decode to?

DevelopmentByDavid commented on app/e2e/tests/wessels-test-setup.ts:
```diff
@@ -0,0 +1,27 @@
+
+    // Login
+    await page.locator('input[type="email"]').click();
+    await page.locator('input[type="email"]').fill('mwess005@ucr.edu');
```
I'd recommend using a fake email address like 123@somewhere.com?  That way this file can be renamed to just test-setup. Unless, is there something I'm missing?

kadey001 replied:
There is already a fake account that the DB is seeded with when running the tests. You can see it under `app/server/prisma/seed.ts`. Email: test@prytaneum.io Pass: password. Please use that account for this setup.

DevelopmentByDavid commented on app/server/src/features/accounts/methods.ts:
```diff
@@ -64,6 +64,36 @@ export async function findOrgsByUserId(userId: string, prisma: PrismaClient) {
+    if (!results) return null;
+
+    // prepare data for graphql layer
+    const userIsModerator = results.moderatorOf.map(({ event }) => event);
+    const userIsInvited = results.invitedOf.map(({ event }) => event);
+    const formattedData = userIsModerator.concat(userIsInvited);
+
+    return formattedData;
```
From my understanding, this could have duplicate events, right?  If someone is both invited and a moderator of a particular event?

Personally, I think the function might be doing too much. It would make sense to me if on the dashboard there's a "events I'm moderating" section. Then this could be split into two functions `getInvitedTo` and `getModeratorOf` with two fields on the graphql model maybe `invitedTo` and `moderatorOf`.

What do you think?

From a user perspective I think it looks more wrong for the same event to appear twice in the same section as opposed to once in one section and once in another section.

kadey001 replied:
I agree with this, it should be split up into separate resolvers and methods. Also the function name here does not describe what it is doing well. It should be something like findInvitedEventsByUserId and findModeratorEventsByUserId. findEventsById is too generic and I think there is another method with the exact same name that could cause confusion.

michael07w replied:
To clarify, would we like to see the events organized on the dashboard first by Ongoing/Upcoming, then by Moderator/Invited? Or only by Moderator/Invited?

DevelopmentByDavid commented on app/server/src/features/accounts/resolvers.ts:
```diff
@@ -3,9 +3,11 @@ import { connectionFromArray, fromGlobalId } from 'graphql-relay';
 import { Resolvers, toGlobalId, runMutation } from '@local/features/utils';
 import { CookieSerializeOptions } from 'fastify-cookie';
 import * as User from './methods';
+import { eventNames } from 'node:process';
```
I don't think this is needed? Probably an accident from pressing tab and vscode auto importing.

DevelopmentByDavid commented on app/e2e/tests/dashboard.spec.ts:
```diff
@@ -0,0 +1,76 @@
+    await Promise.all([
+        page.locator('[aria-label="view live feed of current event"]').click(),
+        page.waitForNavigation(/*{ url: 'http://localhost:8080/events/RXZlbnQ6N2JmZWFmNmItMGJiMi00NTAxLWJjMDYtNWM1ZjFlYWQyZTA1/live'}*/)
+        
```
Extra space here? don't forget to run prettier

DevelopmentByDavid commented on app/e2e/tests/dashboard.spec.ts:
```diff
@@ -0,0 +1,76 @@
+    // Select Live Feed
+    await Promise.all([
+        page.locator('[aria-label="view live feed of current event"]').click(),
+        page.waitForNavigation(/*{ url: 'http://localhost:8080/events/RXZlbnQ6N2JmZWFmNmItMGJiMi00NTAxLWJjMDYtNWM1ZjFlYWQyZTA1/live'}*/)
```
Is the url commented out on purpose?

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,224 @@
+                            <Grid container item direction='column'>
+                                <Grid item xs={12}>
+                                    <List>
+                                        {currentEvents.map(({ node: event }, idx) => {
+                                            if (event?.id && event?.description && event?.title && event?.startDateTime && event?.organization?.name) {
+                                                return (
+                                                    <ListItem
+                                                        button
+                                                        key={event?.id}
+                                                        divider={idx !== currentEvents.length - 1}
+                                                        onClick={handleNav(`/events/${event?.id}/settings`)}
+                                                    >
+                                                        <DashboardEvent
+                                                            key={event.id}
+                                                            id={event.id}
+                                                            title={event.title}
+                                                            description={event.description}
+                                                            startDateTime={event.startDateTime}
+                                                            organization={event.organization.name}
+                                                        />
+                                                        <Link onClick={handleNav(`/events/${event?.id}/live`)}>
+                                                            <Button
+                                                                aria-label='view live feed of current event'
+                                                                variant='contained'
+                                                                color='primary'
+                                                            >
+                                                                Live Feed
+                                                            </Button>
+                                                        </Link>
+                                                    </ListItem>
+                                                );
+                                            }
+                                        })}
```
Considering this logic is duplicated, I think it may be worthwhile to put it into its own component. Maybe `DashboardEventListItem` where you could just do 
```tsx
<List>
  {currentEvents.map((event) =>  <DashboardEventListItem event={event} />}
</List>
```

Also, add any other configurable prop you might need.

DevelopmentByDavid reviewed (changes requested):
Looks good! Thanks for PR. I left some comments.

kadey001 commented:
The dashboard looks good and I like the setup you added. You can now use the file `app/server/prisma/seed.ts` to seed whatever data you need in the DB such as the account used for the tests. However, since the event id will always be unique when creating an event for the tests and the data is time/date related it would be a good idea to create the events you want to test with in the setup alongside the login. That way you can update the urls to be dynamic based on the event id. Let me know if you have any questions about any of this.

kadey001 commented:
> All tests have passed. Ready for Review. #315 ![image](https://user-images.githubusercontent.com/29931619/174917679-8da49c1f-220b-4fc0-9558-722ae8d8a642.png)

The `Dashboard has link to live feed for Current Events.` test fails for me since the static event url used is referencing an event not in the DB for everyone. See my other comment about creating the event as part of the setup to resolve.

michael07w committed:
refactor(frontend): Consolidated styling and removed unnecessary comments

michael07w committed:
refactor(frontend): Consolidated styling and removed unnecessary comments

michael07w committed:
refactor(frontend): Consolidated styling and removed unnecessary comments

michael07w committed:
refactor(frontend): Updated function to get ongoing and upcoming events.
Incorporated useMemo to memoize ongoing and upcoming events. Updated date comparison logic. Formatted code with prettier.

michael07w committed:
refactor(frontend): Updated function to get ongoing and upcoming events.
Incorporated useMemo to memoize ongoing and upcoming events. Updated date comparison logic. Formatted code with prettier.

michael07w committed:
refactor(frontend): Updated function to get ongoing and upcoming events.
Incorporated useMemo to memoize ongoing and upcoming events. Updated date comparison logic. Formatted code with prettier.

michael07w committed:
refactor(frontend): Created new component to limit code duplication in dashboard
Added DashboardEventListItem to avoid code duplication

michael07w committed:
refactor(frontend): Created new component to limit code duplication in dashboard
Added DashboardEventListItem to avoid code duplication

michael07w committed:
refactor(frontend): Created new component to limit code duplication in dashboard
Added DashboardEventListItem to avoid code duplication

michael07w committed:
feat(frontend): Updated event button to match organization button in user's settings

michael07w committed:
feat(frontend): Updated event button to match organization button in user's settings

michael07w committed:
feat(frontend): Updated event button to match organization button in user's settings

dianeshan committed:
Landing tests #303 (#316)

* testing(fullstack): implement landing page tests

* fix(frontend): minor typo fix on landing page

* chore(e2e): moved landing tests to e2e workspace

Co-authored-by: kadey001 <kadey001@ucr.edu>

dianeshan committed:
Landing tests #303 (#316)

* testing(fullstack): implement landing page tests

* fix(frontend): minor typo fix on landing page

* chore(e2e): moved landing tests to e2e workspace

Co-authored-by: kadey001 <kadey001@ucr.edu>

michael07w committed:
test(dashboard): Updated tests to use global test user

michael07w committed:
test(dashboard): Updated tests to use global test user

michael07w committed:
test(dashboard): Updated tests to use global test user

michael07w committed:
test(dashboard): Updated tests to use global test user

michael07w committed:
style(backend): Removed unused import

michael07w committed:
style(backend): Removed unused import

michael07w committed:
style(backend): Removed unused import

michael07w committed:
style(backend): Removed unused import

michael07w commented:
Thank you for all the comments! They were very helpful in pointing me in the right direction.

All comments left above have been addressed in the new commits, with the exception of the following:
https://github.com/c2c-project/prytaneum/pull/243#discussion_r903217765 (moderatorOf & invitedTo)

All dashboard tests still pass.

kadey001 commented on app/e2e/tests/test-setup.ts:
```diff
@@ -0,0 +1,83 @@
+    await page.locator('text=LoginOr, register an account >> button').click();
+    await Promise.all([page.waitForNavigation({ url: 'http://localhost:8080/organizations/me' })]);
+
+    // Add Organization
+    await page.locator('[data-testid="AddIcon"]').click();
+    await page.locator('input[type="text"]').fill('Test Org');
+    await page.locator('button:has-text("Create")').click();
+    await Promise.all([
+        await page.locator('div[role="button"]:has-text("Test Org")').click(),
+        await page.waitForNavigation(),
+    ]);
+
+    // Add Current Event
+    const today = new Date();
+    await addEvent(page, today);
+
+    // Add Upcoming Event
+    const tomorrow = new Date();
+    tomorrow.setDate(tomorrow.getDate() + 1);
+    await addEvent(page, tomorrow);
```
I think it's bad practice to add orgs/events in the global setup. The only thing that should be here is the setup for authentication for the accounts we want to be able to reuse for multiple tests. From what I have read and discussed with David, we think isolating tests as much as possible is important to ensure actions in one test is not causing another to fail/pass, making debugging more complicated and defeating the purpose of e2e testing. For most of the tests we will want to do the setup in the beforeEach/beforeAll function. It is also important to note that tests such as deleting an event cannot re-use authentication because if each test is deleting the same event at the same time with the same account/org only one would/should pass due to conflicts. For a test like that we would want to create an account for each browser and remove them after the tests are completed to ensure the test on each browser is isolated and independent.

DevelopmentByDavid replied:
It would definitely be ideal to setup the data inside of the particular test that needs it or for the suite of tests that need it. I think this is where helper functions can come in handy or we can start using builders to iteratively build up particular types of events.

kadey001 commented on app/e2e/tests/test-teardown.ts:
```diff
@@ -0,0 +1,66 @@
+import { chromium } from '@playwright/test';
+import type { Page } from '@playwright/test';
+
+async function login(page: Page) {
+    await page.locator('input[type="email"]').fill('test@prytaneum.io');
+    await page.locator('input[type="password"]').fill('password');
+    await page.locator('text=LoginOr, register an account >> button').click();
+    await page.waitForURL('http://localhost:8080/organizations/me');
+}
```
I like the idea of helper functions for commonly used actions such as login. Maybe you could move this to a utils file and update it to take in an email + password. Functions for registration, org creation/deletion, and event creation/deletion would also be handy.

DevelopmentByDavid replied:
I think putting this inside a folder named `common/` would work. Then inside of `common/`, we'd have a `setup/` and `interactions/` folder. Setup would contain things that help setup data in the database maybe? Then `interactions` would have common user interactions like `loginWithPassword`.

DevelopmentByDavid replied:
Further expanding on this I think it'd be pretty cool to have tests like so
```ts
import * as I from '@local/common/interactions';
import * as ThereIs from '@local/common/setup';

// ... in some test

// setup
ThereIs.aUserWithEmailAndPassword("email@example.com", "123");

I.navigateTo("/login");
I.loginWithEmailAndPassword("email@example.com", "123");

// insert assertions here
```
I think an API similar to this would be pretty cool / easy to reason about. I'm not sure how hard it is to setup, but I'm just throwing it out there.

kadey001 commented on app/e2e/tests/test-teardown.ts:
```diff
@@ -0,0 +1,66 @@
+    await page.waitForURL('http://localhost:8080/organizations/me');
+    await page.locator('[aria-label="delete organization"]').click();
+    await page.locator('text=Confirm').click();
+}
```
See above comment

kadey001 commented on app/e2e/tests/test-teardown.ts:
```diff
@@ -0,0 +1,66 @@
+    await page.locator(`[aria-label="Enter ${eventText} to delete event"]`).fill(eventText);
+    await page.locator(`[aria-label="Enter ${eventText} again"]`).fill(eventText);
+    await page.locator('button:has-text("Delete Event")').click();
+}
```
^

kadey001 commented on app/e2e/tests/test-teardown.ts:
```diff
@@ -0,0 +1,66 @@
+    // Cleanup
+    await context.close();
+    await browser.close();
+}
```
Not sure all of this is necessary. Any org/event that is created in a test should be cleaned up within that test page in a afterEach or afterAll function. And any data not cleaned up by that should be removed the next time the DB is seeded.

kadey001 commented on app/e2e/tests/dashboard.spec.ts:
```diff
@@ -0,0 +1,87 @@
+import { test, expect } from '@playwright/test';
+
+test('Dashboard has button to create event.', async ({ page }) => {
```
I suggest wrapping tests in describe blocks like the landing tests are. Also might be a good idea to break up dashboard tests into multiple files under a dashboard folder like the landing tests. Playwright can only execute tests in parallel when they are in separate files.

kadey001 commented on app/e2e/tests/test-setup.ts:
```diff
@@ -0,0 +1,83 @@
+
+async function globalSetup() {
+    const browser = await chromium.launch({
+        headless: false,
```
Make sure to update headless to true when you are finished debugging since headless is required when running tests in CI pipeline. Same for the teardown.

DevelopmentByDavid replied:
This should probably be a env variable?

kadey001 replied:
possibly, I think by default it is headless so if you just remove the arg it would be fine as well.

kadey001 committed:
fix(seed): added table drops before seeding

kadey001 committed:
fix(landing e2e): do not use storage state in landing tests

kadey001 committed:
fix(landing e2e): do not use storage state in landing tests

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+import * as React from 'react';
+import { graphql, PreloadedQuery, usePreloadedQuery } from 'react-relay';
+import { useRouter } from 'next/router';
+import { isBefore, isAfter } from 'date-fns';
+
+import { Card, CardContent, Grid, List, Typography, IconButton } from '@mui/material';
+import { makeStyles } from '@mui/styles';
+import { Add } from '@mui/icons-material';
```
Generally, we only try to split up imports if one is local, e.g. `@local/...` and something from a package. Basically, you can remove the newline on line 5.

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+        if (!isLoading && !user) router.push('/');
+    }, [user, router, isLoading]);
+
+    // Store ongoing events
+    const ongoingEvents = React.useMemo(
+        () =>
+            listOfEvents.filter(({ node: event }) => {
+                if (!event.startDateTime || !event.endDateTime) return false;
+                const now = new Date();
+                return isBefore(new Date(event.startDateTime), now) && isAfter(new Date(event.endDateTime), now);
+            }),
+        [listOfEvents]
+    );
+
+    // Store upcoming events
+    const upcomingEvents = React.useMemo(
+        () =>
+            listOfEvents.filter(({ node: event }) => {
+                if (!event.startDateTime) return false;
+                const now = new Date();
+                return isAfter(new Date(event.startDateTime), now);
+            }),
+        [listOfEvents]
+    );
```
Code looks very clean!

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+                                            key={event.id}
+                                            event={event}
+                                            ongoing={true}
+                                            divider={divider}
```
You can get rid of the above and just inline the logic on divider.
```tsx
divider={idx === ongoingEvents.length - 1}
```

You save on code space this way.

```tsx
 {ongoingEvents.map(({ node: event }, idx) => 
  <DashboardEventListItem
      key={event.id}
      event={event}
      ongoing={true}
      divider={idx === ongoingEvents.length - 1}
/>}
```


Also, this could actually be further broken up I think.
```tsx
// Dashboard.tsx
// you can replace this code with the following
<DashboardEventList events={ongoingEvents} emptyStatusText="No ongoing events to display." />

// DashboardEventList.tsx (a file to be created)
function DashboardEventList({ events, emptyStatusText}: props) {
  if (events.length === 0) return <Typography>{emptyStatusText</Typography>
  return (<List>
  // insert other existing code you already have
  </List>);
}

```

It's not quite copy/paste, but I think you get the idea and it would make this whole file really clean I think.

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+                                            key={event.id}
+                                            event={event}
+                                            ongoing={false}
+                                            divider={divider}
```
See previous comment

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+                <Card>
+                    <CardContent style={{ display: 'flex', justifyContent: 'center', padding: 12 }}>
+                        <IconButton aria-label='view future event' onClick={handleNav('/organizations/me')}>
+                            <Add style={{ fontSize: 32, color: 'black' }} />
```
Preferably make a class for this in useStyles -- maybe `addIcon`

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+                        </IconButton>
+                    </CardContent>
+                </Card>
+                <Typography variant='subtitle2' style={{ marginTop: 8 }}>
```
Preferably make a class for this in useStyles.

DevelopmentByDavid commented on app/client/src/features/dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,169 @@
+            </Grid>
+            <Grid item>
+                <Card>
+                    <CardContent style={{ display: 'flex', justifyContent: 'center', padding: 12 }}>
```
Preferrably add a class for this in useStyles.

DevelopmentByDavid commented on app/client/src/features/dashboard/DashboardEventListItem.tsx:
```diff
@@ -0,0 +1,52 @@
+import { DashboardEvent } from '@local/features/dashboard/DashboardEvent';
+
+interface DashboardEventListItemProps {
+    event: {
+        id: string;
+        title: string | null;
+        description: string | null;
+        startDateTime: Date | null;
+        endDateTime: Date | null;
+        organization: {
+            name: string;
+        } | null;
+    };
```
Is event being null special? Is there a reason it can't be just `event?: ...`

DevelopmentByDavid commented on app/client/src/features/dashboard/DashboardEventListItem.tsx:
```diff
@@ -0,0 +1,52 @@
+
+    return (
+        <ListItem key={event.id} divider={divider} button onClick={handleNav(`/events/${event.id}/settings`)}>
+            {event.title && event.description && event.startDateTime && event?.organization?.name && (
```
A helper function would make this more clear IMO.

```tsx
/* Dictates whether or not an event is valid to display to the user */
function isValidEvent(event: Event) {
 return event.title && ...
}
// ...
{isEventValid(event) && ...}
```
This can be defined outside of the react function since it's so simple and doesn't really need to be in the render function.

DevelopmentByDavid commented on app/client/src/features/dashboard/DashboardEventListItem.tsx:
```diff
@@ -0,0 +1,52 @@
+                />
+            )}
+
+            {ongoing && (
+                <ListItemSecondaryAction>
+                    <Button
+                        aria-label='view live feed of current event'
+                        variant='contained'
+                        color='primary'
+                        onClick={handleNav(`/events/${event.id}/live`)}
+                    >
+                        Live Feed
+                    </Button>
+                </ListItemSecondaryAction>
```
I think potentially changing the API of this would help a little with extendability in a general sense.

Maybe instead allow for a children to be passed in so that way the caller can define what type of button to use and the DashboardEventListItem doesn't need to know if it's ongoing or not. So then the caller does
```tsx
<DashboardEventListItem>
  <ListItemSecondaryAction></ListItemSecondaryAction>
</DashboardEventListItem>
```

and then in this file this code just becomes `{children}` which you can destructure from the props.

Then, the caller, which always will know the type of events, will be able to customize the list item as necessary.

DevelopmentByDavid commented on app/e2e/tests/landing/landing-carousel.spec.ts:
```diff
@@ -1,6 +1,7 @@
 import { test, expect } from '@playwright/test';
 
 test.describe('carousel buttons', () => {
+    test.use({ storageState: undefined });
```
What is storageState?

DevelopmentByDavid reviewed (dismissed):
Looks good so far. Thanks for making the previous changes.  

Also, if I made any suggestions that conflict or cause friction with each other, please let me know / feel free to ask in slack or on here.

kadey001 committed:
chore(client): added data-test-id for create org fab

kadey001 committed:
chore(client): added data-test-id for create org fab

kadey001 committed:
fix(e2e): tests use context based on device now, ensuring isolation

kadey001 committed:
fix(e2e): tests use context based on device now, ensuring isolation

kadey001 committed:
chore(e2e): removed unessisary use statement + updated some to run parallel tests

kadey001 committed:
chore(e2e): removed unessisary use statement + updated some to run parallel tests

kadey001 committed:
chore(e2e): added dashboard test with new patern

kadey001 committed:
chore(e2e): added dashboard test with new patern

kadey001 committed:
chore(e2e): removed deprecated dashboard tests and setup/teardown

kadey001 committed:
chore(e2e): removed deprecated dashboard tests and setup/teardown

kadey001 committed:
chore(e2e): added todo + removed debug screenshots

kadey001 committed:
chore(e2e): added todo + removed debug screenshots

kadey001 committed:
chore(e2e): added setup for user-dashboard tests

kadey001 committed:
chore(e2e): added setup for user-dashboard tests

kadey001 committed:
chore(client): removed unsued storageState

kadey001 committed:
chore(client): removed unsued storageState

kadey001 committed:
chore(e2e): removed teardown + increased workers in ci

kadey001 committed:
chore(e2e): removed teardown + increased workers in ci

DevelopmentByDavid commented:
@kadey001 I like the beforeAll and afterAll, should they be beforeEach and afterEach? Or would that not make much sense?

Also, seems some of the tests are failing in CI?

kadey001 commented:
> @kadey001 I like the beforeAll and afterAll, should they be beforeEach and afterEach? Or would that not make much sense?
> 
> Also, seems some of the tests are failing in CI?

I think it just depends. Using beforeEach/afterEach would make the tests even slower and is unnecessary if what they are setting up doesn't change within the scope of the tests. For example if we are testing that events are showing up properly in the dashboard and that they properly redirect you, there would be no benefit to removing the org and event after each test since both don't effect the event they just interact/look at it. I'd say if the test is say editing the event settings then a clean up afterEach would make sense. 

Also that test did pass, it was just considered flaky because it failed and re-ran then passed the next time. It happens sometimes in CI and that's why there is a retry number in the config that is set with the CI env var.

michael07w committed:
Merge branch 'michael07w/load_proper_events_dashboard' of https://github.com/michael07w/prytaneum into michael07w/load_proper_events_dashboard

michael07w committed:
Resolving merge conflicts

michael07w committed:
Resolving merge conflicts

michael07w committed:
test(e2e): implement dashboard tests
This test fails with a high number of workers (>3) on my machine. Pushing to allow others to test locally.

michael07w committed:
test(e2e): implement dashboard tests
This test fails with a high number of workers (>3) on my machine. Pushing to allow others to test locally.

michael07w committed:
test(e2e): Implement Dashboard Tests
Re-implemented dashboard tests using the new fixtures (e.g. dashboardPageOrganizer)

michael07w committed:
test(e2e): Implement Dashboard Tests
Re-implemented dashboard tests using the new fixtures (e.g. dashboardPageOrganizer)

michael07w commented:
I setup all dashboard tests to use the newly-added fixtures.

The PlaywrightDashboardPage can have two contexts: Organizer or User. I am not sure of the differences between the two, so I created all tests in the `organizer-dashboard.spec.ts` file, using the Organizer context option.

michael07w committed:
refactor(frontend): Dashboard
Refactored Dashboard into more granular components to increase readability. Updated backend methods/resolvers to remove any duplicate events that could've been returned to user upon DashboardQuery.

michael07w committed:
refactor(frontend): Dashboard
Refactored Dashboard into more granular components to increase readability. Updated backend methods/resolvers to remove any duplicate events that could've been returned to user upon DashboardQuery.

kadey001 committed:
chore(pkgs): ran yarn install

kadey001 committed:
chore(pkgs): ran yarn install

kadey001 committed:
fix(e2e): moved org tests to test list

kadey001 committed:
fix(e2e): moved org tests to test list

kadey001 commented on app/e2e/tests/dashboard/test.list.ts:
```diff
@@ -0,0 +1,53 @@
+    await organizationsPage.submitOrganizationForm();
+
+    // Create ongoing event
+    const today = new Date();
```
Might be better to set the date in the page class via the constructor so it is set when the page is created and can be used in all necessary methods directly rather than passing it in here for today and tomorrow.

kadey001 commented on app/e2e/tests/dashboard/test.list.ts:
```diff
@@ -0,0 +1,53 @@
+    await organizationsPage.submitEventForm();
+
+    // Create upcoming event
+    const tomorrow = new Date();
```
Refer to the comment above.

kadey001 commented on app/e2e/tests/landing/landing-carousel.spec.ts:
```diff
@@ -0,0 +1,24 @@
+import { test, expect } from '@playwright/test';
+
+test.describe('carousel buttons', () => {
+    test.use({ storageState: undefined });
```
I think this can be removed now that we don't have a single storage state set in the playwright config. This was only necessary before when a login state was re-used across all tests, but now they are only used depending on the page class.

kadey001 commented on app/e2e/tests/landing/landing-login.spec.ts:
```diff
@@ -1,25 +1,23 @@
 
-test.describe('landing page login button', () => {
-    test.use({ storageState: undefined });
+test.describe.configure({ mode: 'parallel' });
```
This can be removed as well since parallel is the default mode.

kadey001 commented on app/e2e/tests/landing/landing-register.spec.ts:
```diff
@@ -1,35 +1,33 @@
 
-test.describe('landing page registration buttons', () => {
-    test.use({ storageState: undefined });
+test.describe.configure({ mode: 'parallel' });
```
Same comment as before

kadey001 commented on app/e2e/tests/test-setup.ts:
```diff
@@ -0,0 +1,17 @@
+import { chromium } from '@playwright/test';
+import { loginWithPassword } from '../common/interactions/loginWithPassword';
```
nitpick but the `..` should probably be replaced with `@local` for consistency.

kadey001 reviewed (changes requested):
Looking good, just a few more changes. Also good work on the refactor.

kadey001 committed:
chore(e2e): removed old test-setup

kadey001 committed:
chore(e2e): removed old test-setup

kadey001 committed:
fix(e2e): serialized dashboard tests

kadey001 committed:
fix(e2e): serialized dashboard tests

kadey001 committed:
ci(workflow): updated artifact to test-results

kadey001 committed:
ci(workflow): updated artifact to test-results

kadey001 committed:
chore(e2e): updated config, reduced to 1 worker

kadey001 committed:
chore(e2e): updated config, reduced to 1 worker

michael07w committed:
chore(e2e): Update Tests to Use Helper Class Variables
Incorporated dates into the helper classes to be used in corresponding tests. Removed unnecessary local configurations in landing tests.

michael07w committed:
chore(e2e): Update Tests to Use Helper Class Variables
Incorporated dates into the helper classes to be used in corresponding tests. Removed unnecessary local configurations in landing tests.

kadey001 committed:
chore(invites): implemented invitedOf list

kadey001 committed:
chore(invites): implemented invitedOf list

michael07w committed:
fix(frontend): Resolved console warning
Resolved console warning regarding nesting divs inside p tags.

michael07w committed:
fix(frontend): Resolved console warning
Resolved console warning regarding nesting divs inside p tags.

kadey001 committed:
Revert "feat(server): pretty print graphql strings in pino"

This reverts commit f4619e4707364bf5eb751a421e9fc621be454657.

michael07w committed:
Merge branch 'michael07w/load_proper_events_dashboard' of https://github.com/michael07w/prytaneum into michael07w/load_proper_events_dashboard

kadey001 committed:
chore(cleanup): removed storage states
